official-antistasi-community/A3-Antistasi
# CS-1465-Hint-Refinement

## What type of PR is this.
* Change

### What have you changed and why?
Information:
* Dismiss hint will now also fire on mouse key-binds. (It used to only fire the EH on keyboard.)
* Dismiss key will fire on key down **only once**. However, there is a performance *boost* to the new implementation used.
  * Fast Xor edge detector executes every frame instead of a more lengthy check previously used.
* RenderLoop pauses when there is nothing to draw. performance *boost*.
* Hints will auto dismiss after 15-sec. 
* 10-sec fade animation for the footer to visualise how much time left.
  * Cost: hint rendered every 15 frames (was 10 sec). Not too bad as it builds on previous optimisations.
  * It does make it more responsive in other situations.

TL;DR More features; More Performance.

### Please specify which Issue this PR Resolves.
closes [#1464](https://github.com/official-antistasi-community/A3-Antistasi/issues/1464)
closes [#1465](https://github.com/official-antistasi-community/A3-Antistasi/issues/1465)

### Please verify the following and ensure all checks are completed.

2. [x] Have you loaded the mission in LAN host?
3. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No (Although independent testing is optimal)
* Yes 

# How can the changes be tested?
> Basic

Make sure mission boots.
Boot messages overwrite each other without a dismiss option.
At the end of init, the dismiss option is enabled.
Create some messages and cycle through them on a low-end PC. OR
Wait for them to auto dismiss after 15-sec. There should be a 10-sec footer fade animation.

> Advanced

Key presses should dismiss **only one** hint until your finger is raised off the key and then put back onto the key.
Assign wild key-binds to User Action 12.
Smash your keyboard.
See if keybindings still work as expected.

> Hint Samples

```sqf
["FooBar", "Hello World"] call A3A_fnc_customHint;

//Save Notify
        ["Restart Notification", "Please make your way to HQ as a restart will occure soon™<br/><br/>Please make sure you save your stats at the Map.", true] remoteExec ["A3A_fnc_customHint", 0, false];

["Vaya...", "Parece que sus notificaciones importantes se cifraron.<br/><br/>Nadie espera el cifrado español.", false, ["Pictures\Intel\laptop_error.paa",1]] remoteExec ["A3A_fnc_customHint", 0, false];

// Pre-parse FooBar(Hello World) NoMacro
        private _iconXML = parseText "<img color='#ffffff' image='functions\UI\images\logo.paa' align='center' size='2' shadow='1' shadowColor='#000000' />";
        private _separator  = parseText "<br/><img color='#e6b24a' image='functions\UI\images\img_line_ca.paa' align='center' size='0.60' />";
        private _header = parseText "<br/><br/><t size='1.2' color='#e5b348' shadow='1' shadowColor='#000000'>FooBar</t>";
        private _body = parseText "<br/><br/><t size='1' color='#ffffff' shadow='1' shadowColor='#000000'>Hello World</t><br/>";
        FooBarParse = composeText [_iconXML, _header, _separator, _body, _separator];
        ["FooBar", FooBarParse] call A3A_fnc_customHint;

```
> Push your PC's hardware.
https://github.com/CalebSerafin/FrostsBites-Random-Arma-Scripts/blob/master/A3A_HintRainbowUberFHD.sqf

### Zip of PBO is Provided
_`c2f211f` - 19Sep2020_
[Antistasi-TEST-CS-1465-Hint-Refinement-2-3-1.Altis.pbo.zip](https://github.com/official-antistasi-community/A3-Antistasi/files/5248690/Antistasi-TEST-CS-1465-Hint-Refinement-2-3-1.Altis.pbo.zip)
